ql/cashflows/overnightindexedcoupon.cpp

```diff
@@ -40,25 +41,34 @@ namespace QuantLib {
                 coupon_ = dynamic_cast<const OvernightIndexedCoupon*>(&coupon);
                 QL_ENSURE(coupon_, "wrong coupon type");
             }
-            Rate swapletRate() const override {
+            Rate swapletRate(const Date& date) const override {
+
+                const Date today = Settings::instance().evaluationDate();
 
-                ext::shared_ptr<OvernightIndex> index =
+                const ext::shared_ptr<OvernightIndex> index =
                     ext::dynamic_pointer_cast<OvernightIndex>(coupon_->index());
 
                 const vector<Date>& fixingDates = coupon_->fixingDates();
+                const vector<Date>& valueDates = coupon_->valueDates();
                 const vector<Time>& dt = coupon_->dt();
 
-                Size n = dt.size(),
-                     i = 0;
+                Size i = 0;
+                auto endDateIterator = valueDates.end();
+                
+                if (date < *endDateIterator) {
```

Comment:
You cannot dereference the end iterator, it's not pointing at an element. It's one beyond the last element.
But anyway, you don't need to check before calling lower_bound.  If the date is beyond the last value date, it will work anyway and return the end iterator.